doc/adr/0015-current-API-authentication-strategy.md

```diff
@@ -0,0 +1,116 @@
+
+This supersedes the sections of [ADR-0005 Wego Core Auth Strategy](./0005-wego-core-auth-strategy.md) not to do with authentication to GitHub/GitLab.
+
+## Context
+
+**Note:** throughout this I refer to "auth methods" and "methods" which, for the sake of this document, mean the general process of authenticating who a request is coming from NOT the specific object method which may be called.
+
+Since ADR-0005 was accepted there have been several changes to Weave GitOps. The largest of these has been the removal of support for making changes to git repositories and a greater use of `impersonation` for authentication to the Kubernetes API.
+
+```mermaid
+flowchart TD
+  start([Request]) --> iterate["Next authentication method (oidc, user etc.)"]
+
+  auth -.-> extAuth["[Optional] external authentication (e.g. OIDC)"]
+  extAuth -.-> auth
+
+  subgraph WithAPIAuthentication
+    iterate --> auth[Authenticate request]
+
+    auth --> isAuth{Is authenticated?}
+
+    isAuth --> |No| isLast{More auth methods?}
+    isAuth --> |Yes| set[Set Principal in context]
+
+  end
+  isAuth --> |Error| unauthorised(["401: unauthorised"])
+  isLast --> |No| unauthorised
+  isLast --> |Yes| auth
+
+  set --> handlers[Handle the request]
+  handlers -.->  k8sCall[Call kubernetes API]
+  k8sCall --> principalType{Principal has token?}
+
+  subgraph ClientConfigWithUser
+    principalType --> |No| impersonate[Call Kubernetes API with impersonation]
+    principalType --> |Yes| passthrough[Call Kubernetes API with principal's token]
+
+  end
+
+  impersonate --> response([Response])
+  passthrough --> response([Response])
+```
+
+### Authentication to the server
+
+Currently there are 3 supported methods:
+
+* OIDC (via header and cookie)
+* Cluster user account
+* Token passthrough
+
+methods must implement the `PrincipalGetter` interface which returns a `UserPrincipal`:
+```go
+type PrincipalGetter interface {
+  Principal(r *http.Request) (*UserPrincipal, error)
+}
+
+type UserPrincipal struct {
+  ID     string   `json:"id"`
+  Groups []string `json:"groups"`
+  Token  string   `json:"-"`
+}
+```
+
+The `PrincipalGetter` is used by the `WithAPIAuthentication` middleware that authenticates requests to the API. Auth methods are enabled via the `--auth-methods` flag. The methods are tested in the following order: cluster-user, token-passthrough then OIDC. The ordering is important as cluster-user and OIDC use the same name for their cookie and OIDC will fail if given the cluster-user cookie. The middleware is used on all calls to the Gitops API that don't match one of the configured `PublicRoutes`. When a request is received it is passed to each enabled method's `Principal` function, the first call to return a non-nil response without erroring succeeds and no further methods are tested. If no methods succeed, or if a method returns an error, the request is rejected with a `401 (unauthorised)` status.
```

Comment:
>is used by the `WithAPIAuthentication` middleware that authenticates requests to the API
where can i find more info about this component?

Reply:
I've linked the code